content/docs/user-guide/machine-learning-pipelines/defining-pipelines.md

```diff
@@ -0,0 +1,264 @@
+We don't want to tell anyone how to write their code or what programs to use!
+However, please be aware that in order to prevent unexpected results when DVC
+reproduces pipeline stages, the underlying code should ideally follow these
+rules:
+
+- Read/write exclusively from/to the specified <abbr>dependencies</abbr> and
+  <abbr>outputs</abbr> (including parameters files, metrics, and plots).
+
+- Completely rewrite outputs. Do not append or edit.
+
+- Stop reading and writing files when the `command` exits.
+
+Also, if your pipeline reproducibility goals include consistent output data, its
+code should be
+[deterministic](https://en.wikipedia.org/wiki/Deterministic_algorithm) (produce
+the same output for any given input): avoid code that increases
+[entropy](https://en.wikipedia.org/wiki/Software_entropy) (e.g. random numbers,
+time functions, hardware dependencies, etc.).
+
+</details>
+
+## Directed acyclic graph (DAG)
+
+DVC represents a pipeline internally as a graph where the nodes are stages and
+the edges are _directed_ dependencies (e.g. A before B). In order for DVC to
+execute the pipeline reliably, its topology should be _acyclic_ -- because
+executing cycles (e.g. A -> B -> C -> A ...) would continue indefinitely. [More
+about DAGs].
+
+Use `dvc dag` to visualize (or export) pipeline DAGs.
+
+[more about dags]: https://en.wikipedia.org/wiki/Directed_acyclic_graph
+
+## `dvc.yaml` metafiles
+
+There are two ways to define <abbr>stages</abbr>. The first and recommended
+method is by writing `dvc.yaml` files directly (`stages` list). The other one is
+via `dvc stage add`, a limited command-line interface to stage setup.
+
+<admon type="tip">
+
+See the full [specification] of `stage` entries.
+
+</admon>
+
+Let's look at a sample `preprocess` stage that runs command
+`python src/preprocess.py`. It depends on the corresponding Python script and on
+a raw data file (ideally already [tracked with DVC]).
+
+```yaml
+stages:
+  preprocess:
+    cmd: python src/preprocess.py
+    deps:
+      - data/raw
+      - src/preprocess.py
+    outs:
+      - data/preprocessed
+```
+
+<admon type="info">
+
+We use [GNU/Linux](https://www.gnu.org/software/software.html) in our examples,
+but Windows or other shells can be used too. Note that while DVC is
+platform-agnostic, the commands you define may have environment-specific
+requirements.
+
+</admon>
+
+You can write the `dvc.yaml` file above directly, or DVC can do it for you with
+the following call:
+
+```dvc
+$ dvc stage add --name preprocess \
+                --deps src/preprocess.py \
+                --deps data/raw \
+                --outs data/preprocessed \
+                python src/preprocess.py
+```
+
+<admon type="tip">
+
+One advantage of using `dvc stage add` is that it will verify the validity of
+the arguments provided (otherwise they won't be checked until execution). A
+disadvantage is that advanced pipelining features such as [templating] are not
+always available this way.
```

Comment:
Great point. Do you think we should add any more detail in this guide to at least hint at what kind of templating and advanced syntax can be used in `dvc.yaml` to make pipelines more flexible?

Reply:
> more detail in this guide to at least hint at what kind of templating and advanced syntax can be used in dvc.yaml

I think that's all covered in https://dvc.org/doc/user-guide/project-structure/dvcyaml-files rn. We could def have a section to list **Advanced features** more specifically I suppose, do you think it's worth it? I expect more traffic in the `dvc.yaml` "guide" (actually a reference I think) that links back to here for further details.